indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java

```diff
@@ -0,0 +1,1021 @@
+      context.getConfiguration().set(ConvertingOutputFormat.PUBLISHED_SEGMENT_KEY, finalSegmentString);
+      context.write(new Text("dataSegment"), new Text(finalSegmentString));
+
+      for (File file : files) {
+        context.getCounter(COUNTER_GROUP, COUNTER_WRITTEN).increment(file.length());
+      }
+      context.progress();
+      context.setStatus("Ready To Commit");
+    }
+
+    // We must be consistent with `io.druid.indexer.HadoopDruidIndexerConfig.makeSegmentOutputPath()`
+    private static Path getOutputBasePathForSegment(
+        DataSegment segment,
+        Path baseOutputPath,
+        Configuration configuration
+    )
+        throws IOException
+    {
+      Path path = new Path(baseOutputPath, segment.getDataSource());
+      if (path.getFileSystem(configuration) instanceof DistributedFileSystem) {
+        path = new Path(
+            path,
+            String.format(
+                "%s_%s",
+                segment.getInterval().getStart().toString(ISODateTimeFormat.basicDateTime()),
+                segment.getInterval().getEnd().toString(ISODateTimeFormat.basicDateTime())
+            )
+        );
+        path = new Path(path, segment.getVersion().replace(':', '_'));
+      } else {
+        path = new Path(
+            path,
+            String.format(
+                "./%s_%s",
+                // Since time contains a `:` we specify this is a relative path or else the URI parser freaks out
+                segment.getInterval().getStart().toString(),
+                segment.getInterval().getEnd().toString()
+            )
+        );
+        path = new Path(path, "./" + segment.getVersion());
+      }
+      return new Path(path, Integer.toString(segment.getShardSpec().getPartitionNum()));
+    }
+
+    @Override
+    protected void setup(Context context) throws IOException, InterruptedException
+    {
+      final File tmpFile = Files.createTempDir();
+      context.getConfiguration().set(TMP_FILE_LOC_KEY, tmpFile.getAbsolutePath());
+    }
+
+    private static URI getURIFromSegment(DataSegment dataSegment)
+    {
+      // There is no good way around this...
+      // TODO: add getURI() to URIDataPuller
+      final Map<String, Object> loadSpec = dataSegment.getLoadSpec();
+      final String type = loadSpec.get("type").toString();
+      final URI segmentLocURI;
+      if ("s3_zip".equals(type)) {
+        segmentLocURI = URI.create(String.format("s3n://%s/%s", loadSpec.get("bucket"), loadSpec.get("key")));
+      } else if ("hdfs".equals(type)) {
+        segmentLocURI = URI.create(loadSpec.get("path").toString());
+      } else if ("local".equals(type)) {
+        try {
+          segmentLocURI = new URI("file", null, loadSpec.get("path").toString(), null, null);
+        }
+        catch (URISyntaxException e) {
+          throw new ISE(e, "Unable to form simple file uri");
+        }
+      } else {
+        try {
+          throw new IAE(
```

Comment:
@fjy as far as I know the hadoop index task has never supported anything that is not available as a filesystem in Hadoop. We should probably fix that and use DataSegmentPushers, but do it in the way that we can leverage native hadoop FS when it's faster than using the generic DataSegmentPusher way, as we currently are.

Doing so will also break backwards compatibility, since we currently rely on credentials in the hadoop config for the hadoop indexing, but rely on credentials in the Druid config for the DataSegmentPushers.

I'll file an issue for that